{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6253f4f4-b025-4fe2-b47e-299f93b8e2a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(suppressWarnings(library(Seurat)))\n",
    "suppressMessages(suppressWarnings(library(tidyverse)))\n",
    "suppressMessages(suppressWarnings(library(ggpubr)))\n",
    "suppressMessages(suppressWarnings(library(ggrepel)))\n",
    "suppressMessages(suppressWarnings(library(ggrastr)))\n",
    "suppressMessages(suppressWarnings(library(RColorBrewer)))\n",
    "suppressMessages(suppressWarnings(library(ggdendro)))\n",
    "suppressMessages(suppressWarnings(library(cowplot)))\n",
    "suppressMessages(suppressWarnings(library(reshape2)))\n",
    "suppressMessages(suppressWarnings(library(gtools)))\n",
    "suppressMessages(suppressWarnings(library(ggplot2)))\n",
    "suppressMessages(suppressWarnings(library(stringr)))\n",
    "suppressMessages(suppressWarnings(library(ggalluvial)))\n",
    "suppressMessages(suppressWarnings(library(openxlsx)))\n",
    "suppressMessages(suppressWarnings(library(lme4)))\n",
    "suppressMessages(suppressWarnings(library(emmeans)))\n",
    "suppressMessages(suppressWarnings(library(dplyr)))\n",
    "suppressMessages(suppressWarnings(library(foreach)))\n",
    "suppressMessages(suppressWarnings(library(doParallel)))\n",
    "suppressMessages(suppressWarnings(library(pheatmap)))\n",
    "suppressMessages(suppressWarnings(library(forcats)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd24c1b0-5ea8-42d3-9f0e-6e31e1d470d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 15, repr.plot.height = 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6b39079-60c1-4369-82e7-532e3af53b55",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1987bb2-f2c6-4c2c-95a3-2e1c57a0c717",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main (Fig. 1-4)\n",
    "neurons <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/Seurat_objs/integrated/neurons_finalized_2025.rds\")\n",
    "glia <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/Seurat_objs/integrated/glia_finalized_2025.rds\")\n",
    "neurons$cell.type <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", neurons$cell.type)\n",
    "glia$cell.type <- gsub(\"_\", \" \", glia$cell.type)\n",
    "DefaultAssay(neurons) <- \"integrated\"\n",
    "DefaultAssay(glia) <- \"integrated\"\n",
    "\n",
    "# Spatial (Fig. 2)\n",
    "spatial <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/analysis/revision/spatial_transcriptomics_dvc/processed_data/spatial_object_preprocessed+labelled.rds\")\n",
    "\n",
    "# CELLEX (Fig. 2-3)\n",
    "cellex_mouse <- read.table(\"/projects/perslab/people/jmg776/projects/DVC/output/CELLEX/output/neurons_mouse_2025.esmu.csv\", header = TRUE, sep = \",\", row.names = 1)\n",
    "cellex_rat <- read.table(\"/projects/perslab/people/jmg776/projects/DVC/output/CELLEX/output/neurons_rat_2025.esmu.csv\", header = TRUE, sep = \",\", row.names = 1)\n",
    "cellex_macaque <- read.table(\"/projects/perslab/people/jmg776/projects/DVC/output/CELLEX/output/neurons_macaque_2025.esmu.csv\", header = TRUE, sep = \",\", row.names = 1)\n",
    "colnames(cellex_mouse) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", colnames(cellex_mouse))\n",
    "colnames(cellex_rat) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", colnames(cellex_rat))\n",
    "colnames(cellex_macaque) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", colnames(cellex_macaque))\n",
    "\n",
    "# Ludwig 2021 (Fig. 3)\n",
    "ludwig2021_neurons <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/Seurat_objs/mouse/mouse_neurons_Ludwig_reintegrated_Seurat_obj.rds\")\n",
    "\n",
    "# Bregma meta (Fig. 3)\n",
    "roi_bregma <- read.table(\"/projects/perslab/people/jmg776/projects/DVC/analysis/revision/spatial_transcriptomics_dvc/ROI_Bregma.tsv\", sep = \"\\t\", header = TRUE)\n",
    "\n",
    "# SCENIC (Fig. 4)\n",
    "rat.AUC <- read.csv(\"/projects/perslab/people/jmg776/projects/DVC/output/SCENIC/rat_DVC_neurons_2025/rat_DVC_neurons_2025_auc.csv\", header = TRUE, row.names = 1)\n",
    "mouse.AUC <- read.csv(\"/projects/perslab/people/jmg776/projects/DVC/output/SCENIC/mouse_DVC_neurons_2025/mouse_DVC_neurons_2025_auc.csv\", header = TRUE, row.names = 1)\n",
    "\n",
    "# DEGs (Fig. 4)\n",
    "DEGs.sc.mouse.neurons.acute <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_mouse_neurons_acute_AM833_2025.rds\")\n",
    "DEGs.sc.mouse.glia.acute <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_mouse_glia_acute_AM833_2025.rds\")\n",
    "DEGs.sc.mouse.neurons.chronic <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_mouse_neurons_chronic_AM833_2025.rds\")\n",
    "DEGs.sc.mouse.glia.chronic <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_mouse_glia_chronic_AM833_2025.rds\")\n",
    "\n",
    "DEGs.sc.rat.neurons.acute <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_rat_neurons_acute_AM833_2025.rds\")\n",
    "DEGs.sc.rat.glia.acute <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_rat_glia_acute_AM833_2025.rds\")\n",
    "DEGs.sc.rat.neurons.chronic <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_rat_neurons_chronic_AM833_2025.rds\")\n",
    "DEGs.sc.rat.glia.chronic <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/single-cell/DESeq2_rat_glia_chronic_AM833_2025.rds\")\n",
    "\n",
    "DEGs.bulk.mouse.acute <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/bulk/DESeq2_mouse_acute.rds\")\n",
    "DEGs.bulk.rat.acute <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/bulk/DESeq2_rat_acute.rds\")\n",
    "DEGs.bulk.mouse.chronic <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/bulk/DESeq2_mouse_chronic.rds\")\n",
    "DEGs.bulk.rat.chronic <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/bulk/DESeq2_rat_chronic.rds\")\n",
    "\n",
    "DEGs.sc.mouse.acute <- c(DEGs.sc.mouse.glia.acute, DEGs.sc.mouse.neurons.acute)\n",
    "DEGs.sc.mouse.chronic <- c(DEGs.sc.mouse.glia.chronic, DEGs.sc.mouse.neurons.chronic)\n",
    "DEGs.sc.rat.acute <- c(DEGs.sc.rat.glia.acute, DEGs.sc.rat.neurons.acute)\n",
    "DEGs.sc.rat.chronic <- c(DEGs.sc.rat.glia.chronic, DEGs.sc.rat.neurons.chronic)\n",
    "\n",
    "names(DEGs.sc.mouse.acute) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", names(DEGs.sc.mouse.acute))\n",
    "names(DEGs.sc.rat.acute) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", names(DEGs.sc.rat.acute))\n",
    "names(DEGs.sc.mouse.chronic) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", names(DEGs.sc.mouse.chronic))\n",
    "names(DEGs.sc.rat.chronic) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", names(DEGs.sc.rat.chronic))\n",
    "\n",
    "# Logistic regression (Fig. 4)\n",
    "glmnet.data <- readRDS(\"/projects/perslab/people/jmg776/projects/DVC/output/DEGs/bulk/glmnet.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d3f06e1-09a6-4bee-8f83-1d201daf100b",
   "metadata": {},
   "source": [
    "# Figure 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb89d050-b07a-4d8c-80d0-22cef053b835",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## B. UMAP of neurons colored by cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fe6caec-e0e8-4aed-a896-15bc485c2d6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_embed_neurons <- as.data.frame(neurons@reductions$umap@cell.embeddings) %>%\n",
    "  mutate(\n",
    "    celltype = neurons$cell.type,\n",
    "    celltype_numeric = as.numeric(factor(celltype, levels = str_sort(unique(celltype), numeric = TRUE))) # Assign numeric IDs to cell types sorted numerically (ensures 'Glu9.0' precedes 'Glu10.0')\n",
    "  )\n",
    "\n",
    "label <- umap_embed_neurons %>%\n",
    "  group_by(celltype_numeric) %>%\n",
    "  summarize(x = median(umap_1), y = median(umap_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ab49e8b-740e-421f-ac26-fe5042c7f3b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(umap_embed_neurons, aes(x = umap_1, y = umap_2, colour = factor(celltype_numeric))) +\n",
    "  geom_point_rast(size = 0.1, alpha = 0.5) +\n",
    "  theme_pubr() +\n",
    "  theme(\n",
    "    axis.line = element_line(colour = \"black\", linewidth = 0.5),\n",
    "    panel.grid = element_blank(),\n",
    "    panel.border = element_blank(),\n",
    "    panel.background = element_blank(),\n",
    "    legend.position = \"none\",\n",
    "    axis.title = element_text(size = 6, face = \"bold\", family = \"sans\"),\n",
    "    axis.text = element_text(size = 6, face = \"bold\", family = \"sans\")\n",
    "  ) +\n",
    "  labs(x = \"UMAP 1\", y = \"UMAP 2\") +\n",
    "  scale_color_manual(values = c(\n",
    "    \"#6D655E\", \"#B45C20\", \"#9ABDA4\", \"#FEEF94\", \"#A6B8B3\",\n",
    "    \"#e586f7\", \"#92AEA5\", \"#76479F\", \"#8F7C00\", \"#8AC48E\",\n",
    "    \"#9D5F35\", \"#F0A0FF\", \"#0075DC\", \"#856249\", \"#FEF495\",\n",
    "    \"#B8228D\", \"#95603C\", \"#565AA7\", \"#EAF09B\", \"#6F94A9\",\n",
    "    \"#C7D69F\", \"#2BCE48\", \"#3A6DAF\", \"#A4BCA3\", \"#D2B3BA\",\n",
    "    \"#4C7AAD\", \"#ED0679\", \"#D80F85\", \"#756458\", \"#E3B8A5\",\n",
    "    \"#ABB5BB\", \"#005C31\", \"#94FFB5\", \"#5EF1F2\", \"#FEE390\",\n",
    "    \"#84C686\", \"#DDB7AC\", \"#AC5D26\", \"#610075\", \"#7FC97F\",\n",
    "    \"#F9BF89\", \"#CDB2C1\", \"#00998F\", \"#5D87AB\", \"#6651A3\",\n",
    "    \"#A55E2E\", \"#FEFA97\", \"#EEBB97\", \"#F4BD90\", \"#C1AFCF\",\n",
    "    \"#A0BAAC\", \"#D9E39D\", \"#4563AC\", \"#BCAED1\", \"#E9BA9E\",\n",
    "    \"#8FC195\", \"#666666\", \"#B1B3C2\", \"#B5C9A1\", \"#FDDE8F\",\n",
    "    \"#FDD88D\", \"#873E9A\", \"#7D6351\", \"#FEE992\", \"#81A1A7\",\n",
    "    \"#f6c7ff\", \"#FDCD8A\", \"#8D6143\", \"#E90E70\", \"#95BF9D\",\n",
    "    \"#C7B0C8\", \"#FDD38B\", \"#C2551D\", \"#FDC286\", \"#B6B1CA\",\n",
    "    \"#FDC788\", \"#FCFD99\", \"#E41667\", \"#E90680\", \"#A72B92\"\n",
    "  )) +\n",
    "  geom_text(\n",
    "    data = label, aes(label = celltype_numeric, x = x, y = y),\n",
    "    size = 6 / .pt,\n",
    "    fontface = \"bold\",\n",
    "    inherit.aes = FALSE\n",
    "  )\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_1b.pdf\", height = 4, width = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c5997d-4bc0-4bec-962e-543e72bf979c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## C. Neuronal cell population proportion bar plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c493d4f-2722-4f2a-8294-16bb419ac6c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_df <- neurons@meta.data %>%\n",
    "  filter(species %in% c(\"macaque\", \"rat\") | dataset == \"mouse.2023\") %>%\n",
    "  mutate(\n",
    "    celltype = cell.type,\n",
    "    celltype_numeric = as.numeric(factor(celltype, levels = str_sort(unique(celltype), numeric = TRUE))),  # Assign numeric IDs to cell types sorted numerically (ensures 'Glu9.0' precedes 'Glu10.0')\n",
    "    celltype_meta = paste0(celltype_numeric, \". \", celltype)  # Converts cell type names to '#. Celltype' to act as UMAP legend\n",
    "  ) %>%\n",
    "  count(species, celltype_numeric, celltype_meta) %>%\n",
    "  group_by(species) %>%\n",
    "  mutate(proportion = n / sum(n)) %>%\n",
    "  ungroup() %>%\n",
    "  arrange(celltype_numeric) %>%\n",
    "  mutate(\n",
    "    celltype_meta = factor(celltype_meta, levels = unique(celltype_meta)),\n",
    "    species = factor(species, levels = c(\"mouse\", \"rat\", \"macaque\"))\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f44fdd7-34b3-4660-8008-1fa2c0005fe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_offsets <- c(\"mouse\" = -0.25, \"rat\" = 0, \"macaque\" = 0.25)\n",
    "full_df$offset <- species_offsets[as.character(full_df$species)]\n",
    "full_df$x_adj <- as.numeric(full_df$celltype_meta) + full_df$offset\n",
    "\n",
    "p <- ggplot(full_df, aes(x = x_adj, y = proportion, fill = species)) +\n",
    "  geom_bar(stat = \"identity\", width = 0.2) +  # Adjust width as needed\n",
    "  scale_x_continuous(\n",
    "    breaks = unique(full_df$celltype_numeric), \n",
    "    labels = levels(full_df$celltype_meta)\n",
    "  ) +\n",
    "  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +  # Display y-axis as percentages\n",
    "  scale_fill_manual(\n",
    "    values = c(\"mouse\" = \"#75C0AF\", \"rat\" = \"#546577\", \"macaque\" = \"#DC7040\")\n",
    "  ) +\n",
    "  theme_minimal() +\n",
    "  labs(x = NULL, y = \"Proportion\", fill = \"Species\") +\n",
    "  theme(\n",
    "    axis.text.x = element_text(\n",
    "      angle = 90, hjust = 1, vjust = 0.5, size = 5\n",
    "    ),\n",
    "    panel.grid = element_blank()\n",
    "  )\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_1c.pdf\", p, width = 15, height = 10)  # Full figure, process further in illustrator"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11a140ee-b6c7-4d4c-8f36-4f58fac74146",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## D. UMAP of species in neurons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e663bfa-2d0e-4203-9e74-79388aeac365",
   "metadata": {},
   "outputs": [],
   "source": [
    "neurons_sub <- subset(neurons, dataset %in% c(\"mouse.2023\", \"rat.2023\", \"macaque.2023\"))\n",
    "\n",
    "umap_embed_neurons <- as.data.frame(neurons_sub@reductions$umap@cell.embeddings) %>%\n",
    "  mutate(\n",
    "    species = factor(neurons_sub$species, levels = c(\"mouse\", \"rat\", \"macaque\"))\n",
    "  ) %>%\n",
    "  sample_frac(1)  # Shuffle the rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8965170d-72a9-479a-b34d-7e04824b42a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(umap_embed_neurons, aes(umap_1, umap_2, colour = species)) +\n",
    "  geom_point_rast(size = 0.1, alpha = 0.5) +\n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(axis.line = element_line(colour = \"black\", size = 0.4), \n",
    "        panel.grid.major = element_blank(),\n",
    "        panel.grid.minor = element_blank(),\n",
    "        panel.border = element_blank(),\n",
    "        panel.background = element_blank(),\n",
    "        legend.position = \"right\",\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        axis.text = element_text(face = \"bold\")) + \n",
    "  labs(x = \"UMAP 1\", y = \"UMAP 2\") + \n",
    "  scale_color_manual(values = c(\"#75C0AF\", \"#546577\", \"#DC7040\"))\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_1d.pdf\", height = 4, width = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2f18aa8-731a-4c7a-81c3-7d6ff5d13712",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## E. UMAP of glial cells colored by species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec828aa2-1cce-4221-9f09-8bf350ccd7b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_embed_glia <- as.data.frame(glia@reductions$umap@cell.embeddings) %>%\n",
    "  mutate(\n",
    "    celltype = glia$cell.type,\n",
    "    species = factor(glia$species, levels = c(\"mouse\", \"rat\", \"macaque\"))\n",
    "  ) %>%\n",
    "  sample_frac(1)\n",
    "\n",
    "label <- umap_embed_glia %>%\n",
    "  group_by(celltype) %>%\n",
    "  summarize(x = median(umap_1), y = median(umap_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e84565-89c9-4b72-b14a-43bf187c7c79",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(umap_embed_glia, aes(x = umap_1, y = umap_2, colour = species)) +\n",
    "  geom_point_rast(size = 0.1, alpha = 0.5) +\n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    axis.line = element_line(colour = \"black\", size = 0.4),\n",
    "    panel.grid = element_blank(),\n",
    "    panel.border = element_blank(),\n",
    "    panel.background = element_blank(),\n",
    "    legend.position = \"none\",\n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    axis.text = element_text(face = \"bold\")\n",
    "  ) +\n",
    "  labs(x = \"UMAP 1\", y = \"UMAP 2\") +\n",
    "  scale_color_manual(values = c(\"#75C0AF\", \"#546577\", \"#DC7040\")) +\n",
    "  geom_text(\n",
    "    data = label, aes(label = celltype, x = x, y = y),\n",
    "    size = 6 / .pt,\n",
    "    fontface = \"bold\",\n",
    "    inherit.aes = FALSE\n",
    "  )\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_1e.pdf\", height = 4, width = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebfa45d6-c21e-4762-a7da-af5b51508307",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Figure 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c119ba-4368-4de6-9533-f71196a85663",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## D. Spatial enrichment plot and Calcr specificity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1da366cb-78ab-4789-b25e-ae5e583af7f3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Enrichment plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae973c1e-7d02-400d-ba9c-924e0309a090",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "calculate_fisher_region_cell_property_enrichment <- function(regions,cell_properties){\n",
    "\n",
    "    region_unique <- unique(na.omit(regions))\n",
    "    cell_property_unique <- unique(na.omit(cell_properties))\n",
    "\n",
    "    enrichment_matrix <- matrix(0, nrow = length(region_unique), ncol = length(cell_property_unique))\n",
    "\n",
    "    rownames(enrichment_matrix) <- sort(region_unique)\n",
    "    colnames(enrichment_matrix) <- sort(cell_property_unique)\n",
    "\n",
    "    for (region in region_unique){\n",
    "        for (cell_property in cell_property_unique){\n",
    "            regions_test <- case_when(regions == region ~ region, TRUE ~ 'Others')\n",
    "            cell_properties_test <- case_when(cell_properties == cell_property ~ cell_property, TRUE ~ 'Others')\n",
    "\n",
    "            enrichment_matrix[region,cell_property] <- table(regions_test,cell_properties_test) %>% \n",
    "                as.data.frame() %>% \n",
    "                arrange(desc(regions_test != \"Others\"),desc(cell_properties_test != \"Others\")) %>% \n",
    "                pivot_wider(names_from = cell_properties_test, values_from = Freq) %>% \n",
    "                dplyr::select(-regions_test) %>% \n",
    "                fisher.test(alternative=\"greater\") %>% \n",
    "                .[[\"p.value\"]]\n",
    "            \n",
    "        }\n",
    "    }\n",
    "\n",
    "    return(enrichment_matrix)\n",
    "}\n",
    "\n",
    "spatial_sub <- subset(spatial, prediction.score.cell.type > 0.6)\n",
    "spatial_sub$region_annotated <- case_when(spatial_sub$region %in% c(\"Sol\",\"Sol_R\",\"Sol_L\") ~ \"NTS\",\n",
    "                                          spatial_sub$region %in% c(\"10N\",\"10N_R\",\"10N_L\") ~ \"DMV\",\n",
    "                                          spatial_sub$region %in% c(\"AP\") ~ \"AP\",\n",
    "                                          spatial_sub$region %in% c(\"12N\") ~ \"Other\", # HYP\n",
    "                                          spatial_sub$region %in% c(\"CC\",\"4V\") ~ \"Other\", # Ventricles\n",
    "                                          spatial_sub$region %in% c(\"Gig\",\"Ret_R\",\"Ret_L\") ~ \"Other\", # RetNuc\n",
    "                                          spatial_sub$region %in% c(\"Pre\") ~ \"Other\", # PrepositusNucleus\n",
    "                                          spatial_sub$region %in% c(\"Gr\",\"Gr_R\",\"Gr_L\") ~ \"Other\", # DorsalColumn\n",
    "                                          spatial_sub$region %in% c(\"CB\") ~ \"Other\") # Cerebellum\n",
    "\n",
    "fisher_matrix <- calculate_fisher_region_cell_property_enrichment(spatial_sub$region_annotated, spatial_sub$predicted.cell.type)\n",
    "colnames(fisher_matrix) <- gsub(\"_(\\\\d+)_\", \"\\\\1.\", colnames(fisher_matrix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bd328dd-ef88-4ded-b31a-b83816ae1b3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "inv_p <- -log(fisher_matrix)\n",
    "inv_p[!is.finite(inv_p)] <- max(inv_p[is.finite(inv_p)], na.rm = TRUE)\n",
    "\n",
    "df <- inv_p %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(\"region\") %>%\n",
    "  pivot_longer(-region, names_to = \"cell_type\", values_to = \"inv_p\") %>%\n",
    "  filter(!cell_type %in% c(\n",
    "    \"Astrocytes\",\"Oligodendrocytes\",\"OPCs\",\"Microglia\",\n",
    "    \"Endothelial_cells\",\"Ependymal_cells\",\"Tanycytes\",\n",
    "    \"Pericytes\",\"VLMCs\",\"Choroid_plexus_cells\"\n",
    "  )) %>%\n",
    "  mutate(\n",
    "    p = exp(-inv_p),\n",
    "    fdr = p.adjust(p, method = \"BH\"),\n",
    "    sig = fdr < 0.05,\n",
    "    cell_type = factor(cell_type, levels = unique(cell_type))\n",
    "  ) %>%\n",
    "  filter(region %in% c(\"DMV\", \"NTS\", \"AP\")) %>%\n",
    "  group_by(cell_type) %>%\n",
    "  filter(any(sig)) %>%\n",
    "  ungroup() %>%\n",
    "  mutate(\n",
    "    cell_type = fct_drop(cell_type),\n",
    "    region    = factor(region, levels = c(\"AP\",\"NTS\",\"DMV\"))\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6888c7b-78dd-4675-ad7c-d05a95a81f46",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(df, aes(region, cell_type)) +\n",
    "  geom_tile(\n",
    "    aes(fill = factor(\n",
    "      if_else(sig, as.character(region), NA_character_),\n",
    "      levels = c(\"AP\",\"NTS\",\"DMV\"))),\n",
    "    color = \"white\", size = 0.2) +\n",
    "  scale_fill_manual(\n",
    "    values = c(\"AP\"=\"#B85283\",\"NTS\"=\"#00A0BA\",\"DMV\"=\"#F0A672\"),\n",
    "    na.value = \"white\", drop = FALSE) +\n",
    "  scale_y_discrete(limits = rev(levels(df$cell_type))) +\n",
    "  theme_minimal() +\n",
    "  theme(\n",
    "    axis.text.x     = element_text(angle = 90, vjust = 0.5, hjust = 1),\n",
    "    axis.title      = element_blank(),\n",
    "    panel.grid      = element_blank(),\n",
    "    legend.position = \"none\")\n",
    "\n",
    "p # Serotonin cluster was on the periphery, not actually in the DVC and will thus be discarded\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_2.1d.pdf\", height = 15, width = 3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "379843d0-cdc3-4732-8961-d72f6529236b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Calcr specificity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4167d22b-a8ee-4b95-bec1-2c426bc05e42",
   "metadata": {},
   "outputs": [],
   "source": [
    "calcr_mouse <- cellex_mouse[\"Calcr\",]\n",
    "rownames(calcr_mouse) <- \"Mouse\"\n",
    "calcr_rat <- cellex_rat[\"Calcr\",]\n",
    "rownames(calcr_rat) <- \"Rat\"\n",
    "calcr_macaque <- cellex_macaque[\"Calcr\",]\n",
    "rownames(calcr_macaque) <- \"Macaque\"\n",
    "\n",
    "calcr_combined <- bind_rows(calcr_mouse, calcr_rat, calcr_macaque)[, c(\"Chat0.0\", \"GABA0.0\", \"GABA0.1\", \"GABA0.2\",\n",
    "                                                                       \"GABA2.0\", \"GABA2.1\", \"GABA2.2\", \"GABA6.1\",\n",
    "                                                                       \"GABA6.2\", \"Glu0.2\", \"Glu2.1\", \"Glu2.2\", \n",
    "                                                                       \"Glu2.3\", \"Glu2.4\", \"Glu3.0\", \"Glu3.1\", \n",
    "                                                                       \"Glu3.2\", \"Glu3.3\", \"Glu3.4\", \"Glu4.0\", \n",
    "                                                                       \"Glu4.1\", \"Glu4.2\", \"Glu4.3\", \"Glu5.0\", \n",
    "                                                                       \"Glu5.1\", \"Glu5.2\", \"Glu5.3\", \"Glu6.0\",\n",
    "                                                                       \"Glu6.1\", \"Glu6.2\", \"Glu8.0\", \"Glu8.1\", \n",
    "                                                                       \"Glu8.2\", \"Glu8.3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e268c0d3-d6c9-470e-a539-e3ce07aa3eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheatmap(\n",
    "  mat = t(calcr_combined),\n",
    "  cluster_rows = FALSE,\n",
    "  cluster_cols = FALSE,\n",
    "  display_numbers = FALSE,\n",
    "  breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1),\n",
    "  color = c(\"#FFFFFF\",\"#91BFDB\",\"#FFFF00\",\"#FC8D59\",\"#D73027\"),\n",
    "  # filename = \"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_2.2d.pdf\",\n",
    "  width = 3,\n",
    "  height = 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c885db43-2dd1-4789-a679-73dd720b6c07",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## G. Marker gene plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "930f54af-2ac5-46b5-8e20-90d5696af9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(neurons) <- neurons$cell.type\n",
    "\n",
    "glu02 <- FindMarkers(neurons, ident.1 = \"Glu0.2\")\n",
    "glu41 <- FindMarkers(neurons, ident.1 = \"Glu4.1\")\n",
    "glu42 <- FindMarkers(neurons, ident.1 = \"Glu4.2\")\n",
    "glu50 <- FindMarkers(neurons, ident.1 = \"Glu5.0\")\n",
    "glu61 <- FindMarkers(neurons, ident.1 = \"Glu6.1\")\n",
    "glu80 <- FindMarkers(neurons, ident.1 = \"Glu8.0\")\n",
    "glu82 <- FindMarkers(neurons, ident.1 = \"Glu8.2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90031f24-8e12-4141-a99f-9db4b4feb090",
   "metadata": {},
   "outputs": [],
   "source": [
    "glu53 %>% arrange(desc(avg_log2FC)) %>% head() # Manually checking and picking genes using pct.1, pct.2 and avglog2FC as guides. Cross referencing with other cell types when plotting..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b5209df-6d73-4f7c-b560-66ad88d33245",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cells <- c(\"Glu0.2\", \"Glu4.1\", \"Glu4.2\", \"Glu5.0\", \"Glu5.3\", \"Glu6.1\", \"Glu8.0\", \"Glu8.2\")\n",
    "genes <- c(\"Shox2\", \"Prrxl1\", \"Olfr78\", \"Etv1\", \"Grp\", \"Cbln2\", \"Dbh\", \"Prlh\")\n",
    "\n",
    "for (i in seq_along(genes)) {\n",
    "\n",
    "  # Create one violin plot for each gene. Will be concatenated later into one plot\n",
    "  gene <- genes[i]\n",
    "  gene_data <- data.frame(expression = neurons@assays$SCT@data[gene, neurons$cell.type %in% selected_cells], \n",
    "                          celltype = factor(neurons$cell.type[neurons$cell.type %in% selected_cells]),\n",
    "                          species = factor(neurons$species[neurons$cell.type %in% selected_cells], levels = c(\"mouse\", \"rat\", \"macaque\")))\n",
    "\n",
    "  violin_plot <- ggplot(gene_data, aes(x = celltype, y = expression, color = celltype, fill = species)) + \n",
    "    geom_violin(scale = \"width\", adjust = 1, show.legend = FALSE, size = 0.2, color = \"black\", width = 0.8) +\n",
    "    theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "    theme(\n",
    "      panel.grid = element_blank(),\n",
    "      axis.line.x = element_line(color = \"black\", size = 0.4),\n",
    "      axis.line.y = element_line(color = \"black\", size = 0.4),\n",
    "      axis.text.y = element_text(face = \"bold\"),\n",
    "      axis.text.x = element_blank(),\n",
    "      axis.ticks.x = element_blank(),\n",
    "      axis.title.y = element_text(face = \"bold.italic\", angle = 360, vjust = 0.5),\n",
    "      plot.margin = unit(c(0, 0, 0, 0), \"cm\"),  # Removed negative bottom margin\n",
    "    ) +\n",
    "    xlab(\"\") + \n",
    "    ylab(gene) + \n",
    "    scale_fill_manual(\n",
    "      values = c(\"#75C0AF\", \"#546577\", \"#DC7040\"), \n",
    "      name = \"\", \n",
    "      labels = c(\"Mouse\", \"Rat\", \"Macaque\")\n",
    "    ) + \n",
    "    scale_y_continuous(breaks = c(0, floor(max(gene_data$expression))),\n",
    "    limits = c(0, NA))\n",
    "\n",
    "  assign(paste0(\"violin_plot\", i), violin_plot)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eef013d4-1fa0-4f27-ad0f-8978ee4b8928",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_plot <- plot_grid(\n",
    "  violin_plot1, violin_plot2, violin_plot3, violin_plot4, violin_plot5,\n",
    "  violin_plot6, violin_plot7, violin_plot8,\n",
    "  align = \"v\", \n",
    "  ncol = 1, \n",
    "  rel_heights = rep(0.8, 10))\n",
    "\n",
    "labels_plot <- ggplot(\n",
    "  data.frame(celltype = factor(selected_cells)),\n",
    "  aes(x = celltype, y = 1)) +\n",
    "  geom_blank() +\n",
    "  theme_void(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    axis.text.x = element_text(face = \"bold\", angle = 45, hjust = 1), ) +\n",
    "  xlab(\"\")\n",
    "\n",
    "p <- plot_grid(marker_plot, labels_plot, ncol = 1, rel_heights = c(10, 1))\n",
    "\n",
    "p # Fix x-axis labels in illustrator\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_2e.pdf\", height = 15, width = 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8e584d7-b1d0-4635-919d-efc4345a001b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Figure 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86017af8-0a80-4dea-8d6f-84441f7fb984",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## A. Dot plot, key genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d283f06-add4-4362-beb1-eebee80ac7a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "format_species <- function(data, species_name) {\n",
    "  expr <- data[genes, ]\n",
    "  expr$gene <- rownames(expr)\n",
    "  expr <- reshape2::melt(expr, id.vars = \"gene\")\n",
    "  colnames(expr) <- c(\"gene\", \"celltype\", \"cellex_score\")\n",
    "  expr$species <- species_name\n",
    "  return(expr)\n",
    "}\n",
    "\n",
    "# Define genes of interest\n",
    "genes <- c(\"Calcr\", \"Ramp3\", \"Glp1r\")\n",
    "\n",
    "# Format data for each species\n",
    "mouse_data <- format_species(cellex_mouse, \"mouse\")\n",
    "rat_data <- format_species(cellex_rat, \"rat\")\n",
    "macaque_data <- format_species(cellex_macaque, \"macaque\")\n",
    "species_data <- rbind(mouse_data, rat_data, macaque_data) %>% filter(gene %in% genes)\n",
    "\n",
    "# Adjust factor levels to ensure the desired plotting order\n",
    "species_data$species <- factor(species_data$species, levels = c(\"macaque\", \"rat\", \"mouse\"))  # Put mouse last for the plot\n",
    "species_data$celltype <- factor(species_data$celltype, levels = levels(factor(neurons$cell.type)))\n",
    "species_data$gene <- factor(species_data$gene, levels = rev(genes))\n",
    "\n",
    "# Adjust the color column with correct factor levels\n",
    "species_data$color <- factor(paste0(species_data$species, round(species_data$cellex_score * 100, 0)),\n",
    "                             levels = c(paste0(\"mouse\", seq(0, 100)),\n",
    "                                        paste0(\"rat\", seq(0, 100)), \n",
    "                                        paste0(\"macaque\", seq(0, 100))))\n",
    "\n",
    "# Define color palettes for each species with named colors\n",
    "palettes <- unlist(list(\n",
    "  mouse = setNames(colorRampPalette(c(\"#C2BEC0\", \"#8AC1B6\", \"#46D3B8\"))(101), paste0(\"mouse\", 0:100)),\n",
    "  rat = setNames(colorRampPalette(c(\"#C2BEC0\", \"#819bb7\", \"#546577\"))(101), paste0(\"rat\", 0:100)),\n",
    "  macaque = setNames(colorRampPalette(c(\"#C2BEC0\", \"#F0A572\", \"#E8682F\"))(101), paste0(\"macaque\", 0:100))\n",
    "))\n",
    "names(palettes) <- sub(\"^[^.]+\\\\.\", \"\", names(palettes))\n",
    "\n",
    "# Create heatmaps for each gene\n",
    "for (i in seq_along(genes)) {\n",
    "  gene <- genes[i]\n",
    "  \n",
    "  heatmap_plot <- ggplot(\n",
    "    subset(species_data[species_data$gene == gene & species_data$celltype %in% c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\"),]), \n",
    "    aes(x = celltype, y = species, color = color)\n",
    "  ) +        \n",
    "    geom_tile(size = 2, color = \"white\", fill = \"grey99\") +\n",
    "    geom_point(size = 2.5, stroke = 0) +\n",
    "    scale_color_manual(values = palettes) +\n",
    "    scale_x_discrete(expand = c(0, 0)) +\n",
    "    xlab(NULL) + ylab(gene) +\n",
    "    theme(\n",
    "      axis.text.y = element_blank(), \n",
    "      axis.title.y = element_text(size = 6, face = \"bold.italic\", angle = 360, vjust = 0.5),\n",
    "      legend.position = \"none\",\n",
    "      axis.line = element_line(colour = \"black\", linewidth = 0.4),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.margin = unit(c(0, 0.1, 0.1, 0.1), \"cm\"),\n",
    "      panel.spacing = unit(0.1, \"cm\"),\n",
    "      panel.background = element_blank(),\n",
    "      panel.grid.minor = element_blank(),\n",
    "      axis.text.x = element_blank()\n",
    "    )\n",
    "  \n",
    "  assign(paste0(\"heatmap_plot\", i), heatmap_plot)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf31d88-1202-41b5-bdb0-06434814db3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "draw_gradient <- function(palette, species_name) {\n",
    "  species_palette <- palette[grep(species_name, names(palette))]  # Extract only the relevant portion of the palette for the species\n",
    "  species_title <- paste(str_to_title(species_name), \"ESμ\")\n",
    "\n",
    "  ggplot(data.frame(\"x\" = seq_along(species_palette),\n",
    "                    color = factor(names(species_palette), levels = names(species_palette))),\n",
    "         aes(x = x, y = 1, fill = color)) +\n",
    "    geom_tile(show.legend = FALSE) +\n",
    "    scale_fill_manual(values = species_palette) +\n",
    "    theme_void() +\n",
    "    ggtitle(species_title) +\n",
    "    scale_x_continuous(breaks = seq(0, 100, by = 25),\n",
    "                       labels = seq(0, 1, by = 0.25)) +\n",
    "    theme(axis.text.x = element_text(size = 6, face = \"bold\"),\n",
    "          plot.title = element_text(hjust = 0.5, size = 6, face = \"bold\"))\n",
    "}\n",
    "\n",
    "labels_plot <- ggplot(data.frame(celltype = factor(c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\"))), \n",
    "                      aes(x = celltype, y = 1)) +\n",
    "  geom_blank() +\n",
    "  theme_void() +\n",
    "  theme(axis.text.x = element_text(size = 10, face = \"bold\", angle = 45, hjust = 1),\n",
    "        axis.ticks.x = element_line()) +\n",
    "  xlab(\"\")\n",
    "\n",
    "legend_plot <- plot_grid(\n",
    "  draw_gradient(palettes, \"mouse\"), \n",
    "  draw_gradient(palettes, \"rat\"), \n",
    "  draw_gradient(palettes, \"macaque\"), \n",
    "  ncol = 1\n",
    ")\n",
    "\n",
    "p <- plot_grid(legend_plot,\n",
    "               heatmap_plot1, heatmap_plot2, heatmap_plot3,\n",
    "               labels_plot, \n",
    "               align = \"v\", \n",
    "               ncol = 1)\n",
    "\n",
    "p # Fix x-axis label in illustrator\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_3a.pdf\", height = 8, width = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac01473-5528-45a1-b2df-0b6af69a3c25",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## C. Bregma levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "847ba2c8-5e12-47ad-b3ca-9f0a1b807bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial@meta.data %>% filter(prediction.score.cell.type > 0.6) %>% select(c(\"section_index\", \"run_index\", \"predicted.cell.type\")) %>% \n",
    "  filter(!predicted.cell.type %in% c(\"Astrocytes\", \"Oligodendrocytes\",\"OPCs\",\"Microglia\",\"Endothelial_cells\",\"Ependymal_cells\",\"Tanycytes\",\"Pericytes\",\"VLMCs\", \"Choroid_plexus_cells\")) %>%\n",
    "  filter(!(section_index %in% c(\"D1\",\"D2\") & run_index == \"spatial1\")) %>%\n",
    "  group_by(run_index, section_index, predicted.cell.type) %>%\n",
    "  summarize(counts = n()) -> cell_type_by_section\n",
    "cell_type_by_section <- cell_type_by_section %>% ungroup() %>% group_by(predicted.cell.type) %>% mutate(fraction = counts / sum(counts))\n",
    "\n",
    "roi_bregma$Run <- paste0(\"spatial\", roi_bregma$Run - 1)\n",
    "roi_bregma <- roi_bregma %>% select(Run, Well, Bregma)\n",
    "roi_bregma[roi_bregma$Run == \"spatial2\" & roi_bregma$Well == \"C2\", \"Bregma\"] <- -13.71\n",
    "cell_type_by_section <- cell_type_by_section %>% left_join(roi_bregma, by = c(\"run_index\" = \"Run\", \"section_index\" = \"Well\"))\n",
    "cell_type_by_section$predicted.cell.type <- gsub(\"^(\\\\w+)_(\\\\d{1,2})_(\\\\d)$\", \"\\\\1\\\\2.\\\\3\", cell_type_by_section$predicted.cell.type)\n",
    "cell_type_by_section$Bregma <- gsub(\",\", \".\", cell_type_by_section$Bregma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3386d2bb-70b0-4d0c-8fcb-2e34a5b09526",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- cell_type_by_section %>% filter(predicted.cell.type %in% c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\")) %>% ggplot() + geom_col(aes(x = fraction, y = factor(Bregma,levels = sort(unique(cell_type_by_section$Bregma), decreasing = TRUE)))) + \n",
    "                                            facet_grid(~predicted.cell.type) + \n",
    "                                            scale_x_continuous(breaks = c(0,1), limits = c(0,1)) + \n",
    "                                            theme(strip.text.x = element_text(angle = 90))\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_3c.pdf\", height = 5, width = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddecf5f5-2880-423c-a353-b15fe0c7cc3b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## E. Sankey diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f8d156-da88-4bca-9e7a-d2fe2233fa1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ludwig2021_neurons <- subset(ludwig2021_neurons, cells = which(colnames(ludwig2021_neurons) %in% gsub(\"_[0-9]+$\", \"\", colnames(neurons)[which(neurons$year == \"2021\")])))\n",
    "neurons$ludwig.celltype <- NA\n",
    "neurons$ludwig.celltype[na.omit(match(colnames(ludwig2021_neurons), gsub(\"_[0-9]+$\", \"\", colnames(neurons))))] <- ludwig2021_neurons$cell.subtype2\n",
    "\n",
    "celltype_data <- subset(\n",
    "  data.frame(\n",
    "    celltype = neurons$cell.type[neurons$dataset == \"mouse.2021\"],\n",
    "    ludwig_celltype = neurons$ludwig.celltype[neurons$dataset == \"mouse.2021\"]),\n",
    "  !is.na(ludwig_celltype))\n",
    "\n",
    "for (ludwig_type in unique(celltype_data$ludwig_celltype)) {\n",
    "  freq_table <- as.data.frame(table(celltype_data$celltype[celltype_data$ludwig_celltype == ludwig_type]))  # Check how many of present dataset map to current Ludwig2021 cell type\n",
    "  unmappable <- freq_table$Var1[freq_table$Freq < (sum(freq_table$Freq) * 0.2)]  # If less than 20% of a given cell type map to Ludwig2021, it will be considered unmapped\n",
    "  \n",
    "  celltype_data$ludwig_celltype[celltype_data$celltype %in% unmappable & celltype_data$ludwig_celltype == ludwig_type] <- \"unmapped\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78e9d6b8-2dc9-48fd-9fc5-0727e9c570e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts <- celltype_data %>%  # Aggregate counts for plotting\n",
    "  group_by(celltype, ludwig_celltype) %>%\n",
    "  summarise(Freq = n(), .groups = 'drop') %>%\n",
    "  filter(celltype %in% c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\"))  # Only interested in subset\n",
    "\n",
    "p <- ggplot(counts, aes(axis1 = celltype, axis2 = ludwig_celltype, y = Freq)) +\n",
    "  geom_alluvium(aes(fill = celltype), width = 1/12) +\n",
    "  geom_stratum(width = 1/12, fill = \"grey\", color = \"black\") +\n",
    "  geom_text(\n",
    "    stat = \"stratum\",\n",
    "    aes(label = after_stat(stratum),\n",
    "      x = after_stat(x) + 0.1 * ifelse(after_stat(x) == 1, -1, 1),\n",
    "      hjust = ifelse(after_stat(x) == 1, 1, 0)),\n",
    "      size = 3) +\n",
    "  scale_x_discrete(limits = c(\"Present\", \"Ludwig et. al 2021\")) +\n",
    "  theme_minimal() +\n",
    "  theme(legend.position = \"none\",\n",
    "    axis.title = element_blank(),\n",
    "    axis.text.x = element_text(size = 12),\n",
    "    axis.text = element_blank(),\n",
    "    axis.ticks = element_blank(),\n",
    "    panel.grid = element_blank(),\n",
    "    plot.title = element_blank())\n",
    "\n",
    "p # Undecided on colors, will adjust later in illustrator\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_3e.pdf\", height = 5, width = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43010557-392c-4435-af01-f221a0c66f4e",
   "metadata": {},
   "source": [
    "# Figure 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cf94e3c-14cb-4704-8f54-0137c37d7508",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## A. SCENIC analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "886c4ea0-fdb9-47a5-b99b-3304ca005964",
   "metadata": {},
   "outputs": [],
   "source": [
    "identify_da_regulons <- function(seurat_obj, celltypes, group1, group2) {\n",
    "  lmer_results <- foreach(celltype = celltypes) %dopar% {\n",
    "    \n",
    "    seurat_obj_sub <- subset(seurat_obj, cell.type == celltype & treatment %in% c(group2, group1))\n",
    "    sample_ids <- unique(seurat_obj_sub$hash.ID)\n",
    "    \n",
    "    regulons <- rownames(seurat_obj_sub@assays$regulon)[\n",
    "      which(apply(seurat_obj_sub@assays$regulon@data, 1, function(x) {\n",
    "        sum(x != 0) >= (ncol(seurat_obj_sub) * 0.20)\n",
    "      }))\n",
    "    ]\n",
    "    \n",
    "    regulons <- \"Fos...\"  # Placeholder to be replaced by actual regulons selection criteria\n",
    "    \n",
    "    # Prepare data for modeling\n",
    "    data <- cbind.data.frame(\n",
    "      treatment = as.factor(seurat_obj_sub$treatment), \n",
    "      sample = as.factor(seurat_obj_sub$hash.ID),\n",
    "      pool = as.factor(seurat_obj_sub$pool),\n",
    "      run = as.factor(seurat_obj_sub$run),\n",
    "      calcr = seurat_obj_sub@assays$SCT@data[\"Calcr\",],\n",
    "      treatment = factor(seurat_obj_sub$treatment, levels = c(group2, group1))\n",
    "    )\n",
    "    \n",
    "    results <- data.frame(matrix(NA, nrow = length(regulons), ncol = 4))\n",
    "    colnames(results) <- c(\"regulon\", \"beta\", \"SE\", \"p_value\")\n",
    "    \n",
    "    for (i in seq_along(regulons)) {\n",
    "      set.seed(i)\n",
    "      data$regulon <- seurat_obj_sub@assays$regulon@data[regulons[i],]\n",
    "      \n",
    "      # Linear mixed effects model\n",
    "      model <- lmer(\n",
    "        regulon ~ treatment + (1 | sample),\n",
    "        data = data,\n",
    "        REML = TRUE\n",
    "      )\n",
    "      \n",
    "      emm <- lsmeans(model, pairwise ~ treatment, adjust = NULL)\n",
    "      emm_summary <- summary(emm$contrasts)\n",
    "      \n",
    "      results$regulon[i] <- regulons[i]\n",
    "      contrast <- paste0(\"(\", group1, \") - (\", group2, \")\")\n",
    "      results$p_value[i] <- emm_summary$p.value[which(emm_summary$contrast == contrast)]\n",
    "      results$SE[i] <- emm_summary$SE[which(emm_summary$contrast == contrast)]\n",
    "      results$beta[i] <- emm_summary$estimate[which(emm_summary$contrast == contrast)]\n",
    "    }\n",
    "    \n",
    "    results$celltype <- celltype\n",
    "    results\n",
    "  }\n",
    "  \n",
    "  names(lmer_results) <- celltypes\n",
    "  return(lmer_results)\n",
    "}\n",
    "\n",
    "process_SCENIC_data <- function(species) {\n",
    "  # Set parameters based on species\n",
    "  if (species == \"mouse\") {\n",
    "    seurat_obj <- mouse_neurons\n",
    "    comparison_acute <- \"Cagrilintide vs. vehicle (4 hours)\"\n",
    "    comparison_chronic <- \"Cagrilintide vs. weight-matched control (8 days)\"\n",
    "  } else if (species == \"rat\") {\n",
    "    seurat_obj <- rat_neurons\n",
    "    comparison_acute <- \"Cagrilintide vs. vehicle control (4 hours)\"\n",
    "    comparison_chronic <- \"Cagrilintide vs. weight-matched control (8 days)\"\n",
    "  }\n",
    "  \n",
    "  # Process acute data\n",
    "  lmer_acute <- identify_da_regulons(\n",
    "    seurat_obj = seurat_obj,\n",
    "    celltypes = c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\"),\n",
    "    group1 = \"A8-A\",\n",
    "    group2 = \"V-A\"\n",
    "  )\n",
    "  lmer_acute <- dplyr::bind_rows(lmer_acute, .id = \"celltype\")\n",
    "  lmer_acute$p_adj <- p.adjust(lmer_acute$p_value, method = \"fdr\")\n",
    "  lmer_acute$comparison <- comparison_acute\n",
    "  \n",
    "  # Process chronic data\n",
    "  lmer_chronic <- identify_da_regulons(\n",
    "    seurat_obj = seurat_obj,\n",
    "    celltypes = c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\"),\n",
    "    group1 = \"A8-C\",\n",
    "    group2 = \"WM-C\"\n",
    "  )\n",
    "  lmer_chronic <- dplyr::bind_rows(lmer_chronic, .id = \"celltype\")\n",
    "  lmer_chronic$p_adj <- p.adjust(lmer_chronic$p_value, method = \"fdr\")\n",
    "  lmer_chronic$comparison <- comparison_chronic\n",
    "  \n",
    "  # Combine acute and chronic data\n",
    "  lmer_combined <- rbind(lmer_acute, lmer_chronic)\n",
    "  lmer_combined$celltype <- factor(\n",
    "    lmer_combined$celltype,\n",
    "    levels = c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\")\n",
    "  )\n",
    "  \n",
    "  # Add significance symbols\n",
    "  lmer_combined$p_symbol <- \"\"\n",
    "  lmer_combined$p_symbol[lmer_combined$p_adj < 0.05] <- \"*\"\n",
    "  lmer_combined$p_symbol[lmer_combined$p_adj < 0.01] <- \"**\"\n",
    "  lmer_combined$p_symbol[lmer_combined$p_adj < 0.001] <- \"***\"\n",
    "  \n",
    "  # Add species column\n",
    "  lmer_combined$species <- species\n",
    "  \n",
    "  # Extract time point from comparison\n",
    "  lmer_combined$time_point <- ifelse(grepl(\"4 hours\", lmer_combined$comparison), \"4 hours\", \"8 days\")\n",
    "  \n",
    "  return(lmer_combined)\n",
    "}\n",
    "\n",
    "mouse_neurons <- subset(neurons, dataset == \"mouse.2023\")\n",
    "mouse_neurons[[\"regulon\"]] <- CreateAssayObject(data = t(as.matrix(mouse.AUC)))\n",
    "rat_neurons <- subset(neurons, species == \"rat\")\n",
    "rat_neurons[[\"regulon\"]] <- CreateAssayObject(data = t(as.matrix(rat.AUC)))\n",
    "\n",
    "lmer_mouse <- process_SCENIC_data(\"mouse\")\n",
    "lmer_rat <- process_SCENIC_data(\"rat\")\n",
    "lmer_combined <- rbind(lmer_mouse, lmer_rat) %>%\n",
    "  mutate(\n",
    "    x = as.numeric(celltype),\n",
    "    x = ifelse(species == \"mouse\", x - 0.2, x + 0.2)  # Small offset for subsequent plotting\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52161a29-8d51-442d-8235-19ec5740b804",
   "metadata": {},
   "outputs": [],
   "source": [
    "signif_data <- lmer_combined %>%  # Adding significance symbols to plot\n",
    "  filter(p_symbol != \"\") %>%\n",
    "  mutate(y = beta + SE + 0.001)\n",
    "\n",
    "p <- ggplot(lmer_combined, aes(x = celltype, y = beta, color = species, group = species)) +\n",
    "  geom_point(size = 1, position = position_dodge(width = 0.6)) +\n",
    "  geom_errorbar(\n",
    "    aes(ymin = beta - SE, ymax = beta + SE),\n",
    "    width = 0.1,\n",
    "    position = position_dodge(width = 0.6)\n",
    "  ) +\n",
    "  geom_hline(yintercept = 0, linetype = \"dashed\", color = \"black\", size = 0.1) +\n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1),\n",
    "    legend.position = \"top\",  # Place legend above the plot\n",
    "    plot.margin = unit(c(0, 0, 0, 0), \"cm\"),\n",
    "    legend.title = element_blank()\n",
    "  ) +\n",
    "  ylab(expression(bold(beta))) +\n",
    "  xlab(\"\") +\n",
    "  scale_color_manual(values = c(\"#75C0AF\", \"#546577\"), labels = c(\"Mouse\", \"Rat\")) +\n",
    "  facet_wrap(~time_point) +\n",
    "  geom_text(\n",
    "    data = signif_data,\n",
    "    aes(label = p_symbol, y = y),\n",
    "    size = 2,\n",
    "    fontface = \"bold\",\n",
    "    color = \"black\",\n",
    "    position = position_dodge(width = 0.6)\n",
    "  )\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4a.pdf\", height = 4, width = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab9addc4-06fb-4211-9993-9a9be3bdc3b3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## C. DEGs treatment vs. control after 4 hours and 8 days (bulk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b28f9bb8-a8bb-4247-88ce-6d1f6ef811a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "DEGs <- data.frame(\n",
    "  treatment = factor(c(\"Acute\", \"Subchronic\", \"Acute\", \"Subchronic\")),\n",
    "  species = factor(c(\"mouse\", \"mouse\", \"rat\", \"rat\"), levels = c(\"mouse\", \"rat\")),\n",
    "  sign_genes = c(  # Summarizing all significant genes across species and conditions\n",
    "    sum(DEGs.bulk.mouse.acute$padj < 0.05, na.rm = TRUE),\n",
    "    sum(DEGs.bulk.mouse.chronic$padj < 0.05, na.rm = TRUE),\n",
    "    sum(DEGs.bulk.rat.acute$padj < 0.05, na.rm = TRUE),\n",
    "    sum(DEGs.bulk.rat.chronic$padj < 0.05, na.rm = TRUE)\n",
    "  )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f6846bf-e170-417f-b322-59e96abc488e",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(DEGs, aes(x = treatment, y = sign_genes, fill = species)) +\n",
    "  geom_bar(stat = \"identity\", width = 0.6, position = \"dodge\", size = 0.3, color = \"black\") +\n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    legend.position = \"top\",\n",
    "    text = element_text(face = \"bold\"),\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1),\n",
    "    axis.line = element_line(color = \"black\", size = 0.3),\n",
    "    axis.ticks = element_line(color = \"black\", size = 0.3),\n",
    "    legend.key.height = unit(1, \"mm\"),\n",
    "    legend.key.width = unit(2, \"mm\"),\n",
    "    legend.title = element_blank()\n",
    "  ) +\n",
    "  labs(y = \"Differentially expressed genes\") +\n",
    "  scale_fill_manual(values = c(\"#75C0AF\", \"#546577\"), labels = c(\"Mouse\", \"Rat\")) +\n",
    "  scale_x_discrete(labels = c(\"4 hours\", \"8 days\"))\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4c.pdf\", height = 4, width = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0a6e324-84e9-4d05-b904-384cf85d26ac",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## D. Accuracy of regularized logistic regression classifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04ad24c5-a10c-4c92-9a43-15c0371debf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "glmnet_data <- glmnet.data %>%\n",
    "    mutate(\n",
    "        accuracy = accuracy * 100,\n",
    "        species = gsub(\".acute|.chronic\", \"\", rownames(.)),\n",
    "        study = gsub(\"mouse.|rat.\", \"\", rownames(.))\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edcf315d-3d98-48b7-b62c-41a0bb14f43b",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(glmnet_data, aes(x = study, y = accuracy, fill = species)) +\n",
    "  geom_bar(stat = \"identity\", width = 0.5, position = \"dodge\", size = 0.3, color = \"black\") +\n",
    "  theme_classic(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    legend.position = \"top\",\n",
    "    text = element_text(face = \"bold\"),\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1),\n",
    "    axis.line = element_line(color = \"black\", size = 0.3),\n",
    "    axis.ticks = element_line(color = \"black\", size = 0.3),\n",
    "    legend.key.height = unit(1, \"mm\"),\n",
    "    legend.key.width = unit(2, \"mm\"),\n",
    "    legend.title = element_blank()\n",
    "  ) +\n",
    "  labs(y = \"Accuracy (%)\") +\n",
    "  scale_fill_manual(values = c(\"#75C0AF\", \"#546577\"), labels = c(\"Mouse\", \"Rat\")) +\n",
    "  scale_x_discrete(labels = c(\"4 hours\", \"8 days\")) +\n",
    "  scale_y_continuous(limits = c(0, 100), breaks = seq(0, 100, 25))\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4d.pdf\", height = 4, width = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96513b91-86b4-47bc-8f35-49112776f9dc",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## E. DEGs treatment vs. control after 4 hours and 8 days (single-cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24de1861-49b7-4953-83b9-8a663b82ea55",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_DEGs_data <- function(celltypes, DEGs_sc_rat, DEGs_sc_mouse, comparison_label) {\n",
    "  DEGs_data <- data.frame(celltype = celltypes, mouse = NA, rat = NA, stringsAsFactors = FALSE)\n",
    "  \n",
    "  for (i in seq_along(celltypes)) {\n",
    "    celltype <- celltypes[i]\n",
    "    \n",
    "    # Rat DEGs\n",
    "    if (!is.null(DEGs_sc_rat[[celltype]])) {\n",
    "      DEGs_data$rat[i] <- sum(DEGs_sc_rat[[celltype]]$padj < 0.05, na.rm = TRUE)\n",
    "    }\n",
    "    \n",
    "    # Mouse DEGs\n",
    "    if (!is.null(DEGs_sc_mouse[[celltype]])) {\n",
    "      DEGs_data$mouse[i] <- sum(DEGs_sc_mouse[[celltype]]$padj < 0.05, na.rm = TRUE)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  DEGs_data$comparison <- comparison_label\n",
    "  return(DEGs_data)\n",
    "}\n",
    "\n",
    "celltypes <- c(\"Glu4.2\", \"Glu8.0\", \"Glu8.2\")\n",
    "\n",
    "# Acute DEGs\n",
    "DEGs_sc_acute <- get_DEGs_data(\n",
    "  celltypes,\n",
    "  DEGs.sc.rat.acute,\n",
    "  DEGs.sc.mouse.acute,\n",
    "  \"Cagrilintide vs. vehicle control (4 hours)\"\n",
    ")\n",
    "\n",
    "# Chronic DEGs\n",
    "DEGs_sc_chronic <- get_DEGs_data(\n",
    "  celltypes,\n",
    "  DEGs.sc.rat.chronic,\n",
    "  DEGs.sc.mouse.chronic,\n",
    "  \"Cagrilintide vs. weight-matched control (8 days)\"\n",
    ")\n",
    "\n",
    "# Combine and format\n",
    "DEGs <- rbind(DEGs_sc_acute, DEGs_sc_chronic)\n",
    "DEGs_melt <- melt(DEGs, id.vars = c(\"celltype\", \"comparison\"))\n",
    "colnames(DEGs_melt) <- c(\"celltype\", \"comparison\", \"species\", \"genes\")\n",
    "DEGs_melt$genes[is.na(DEGs_melt$genes)] <- 0  # One instance where DEGs is NA, need to account for that"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87f3d7ac-e7a4-40e7-8935-87a2351568cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(DEGs_melt, aes(x = celltype, y = genes, fill = species)) +\n",
    "  geom_bar(position = position_dodge(), stat = \"identity\", color = \"black\", width = 0.8) +\n",
    "  facet_wrap(~comparison, ncol = 1) +\n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1, face = \"bold\"),\n",
    "    axis.text.y = element_text(face = \"bold\"),\n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    legend.title = element_blank(),\n",
    "    strip.text = element_text(face = \"bold\")\n",
    "  ) +\n",
    "  xlab(NULL) +\n",
    "  ylab(\"Differentially Expressed Genes\") +\n",
    "  scale_y_continuous(limits = c(0, max(15, max(DEGs_melt$genes, na.rm = TRUE)))) +\n",
    "  scale_fill_manual(values = c(\"#75C0AF\", \"#546577\"))\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4e.pdf\", width = 5, height = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c18b803-d2f7-419d-9ee1-d4beb68f0a99",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## F. Volcano (bulk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70161238-e196-4681-8df7-cdbb87a5a03f",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano <- DEGs.bulk.rat.chronic %>%\n",
    "  filter(!is.na(padj)) %>%\n",
    "  mutate(\n",
    "    label = if_else(padj == min(padj), rownames(.), \"\"),  # Adds label to the smallest adj. pvalue for subsequent plotting\n",
    "    col = case_when(\n",
    "      padj <= 0.05 & abs(log2FoldChange) >= 0.5 ~ 1,  # Setting color levels based on numeric value\n",
    "      padj <= 0.05 & abs(log2FoldChange) < 0.5  ~ 2,\n",
    "      padj > 0.05 & abs(log2FoldChange) >= 0.5  ~ 3,\n",
    "      TRUE                                      ~ 4\n",
    "    )\n",
    "  ) %>%\n",
    "  select(log2FoldChange, padj, col, label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c356db94-787b-439c-aabd-e8ba1236efd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(volcano, aes(y = -log10(padj), x = log2FoldChange,\n",
    "                                    fill = factor(col), label = label)) +\n",
    "  geom_point(shape = 21, size = 3, alpha = 1, stroke = 0.2) + \n",
    "  geom_hline(yintercept = -log10(0.05), linetype = \"dashed\") +\n",
    "  geom_vline(xintercept = c(-0.5, 0.5), linetype = \"dashed\") + \n",
    "  geom_text_repel(fontface = \"bold.italic\", size = 2, max.overlaps = 20) + \n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(legend.position = \"none\",\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        axis.text = element_text(face = \"bold\"),\n",
    "        axis.line = element_line(size = 0.4),\n",
    "        plot.margin = margin(0, 0, 0, 0, \"cm\")) +\n",
    "  xlab(expression(bold(Log[2] * \" fold-change\"))) + \n",
    "  ylab(expression(bold(-log[10] * \"(\" * italic(P) * \")\"))) + \n",
    "  scale_fill_manual(values = c(\"1\" = \"#95567D\", \"2\" = \"#F9D4EC\", \n",
    "                               \"3\" = \"grey70\", \"4\" = \"grey90\")) +\n",
    "  xlim(c(-2.5, 2.5))  # Cuts off 9 points in total, but these points are not interesting\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4f.pdf\", width = 5, height = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec714263-7b7a-4797-8179-1265cf634727",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## G. Prlh feature plot (single-cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "757ecce9-38e0-47d1-b0c4-60c34136e4c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_neurons <- subset(neurons, dataset == \"rat.2023\")\n",
    "DefaultAssay(rat_neurons) <- \"RNA\"\n",
    "\n",
    "rat_prlh_cells <- WhichCells(rat_neurons, expression = Prlh > 5)\n",
    "\n",
    "umap_embed_rat_neurons <- neurons@reductions$umap@cell.embeddings %>%\n",
    "  as.data.frame() %>%\n",
    "  mutate(\n",
    "    species = neurons$species,\n",
    "    celltype = neurons$cell.type,\n",
    "    Prlh_expressed = rownames(.) %in% rat_prlh_cells,  # Label cells where Prlh is expressed\n",
    "    color_group = ifelse(Prlh_expressed, as.character(species), \"Not expressed\"),\n",
    "    color_group = factor(color_group, levels = c(\"rat\", \"Not expressed\"))\n",
    "  ) %>%\n",
    "  sample_frac(1) %>%  # Shuffle rows\n",
    "  arrange(desc(color_group == \"Not expressed\"))  # Ensure background is plotted first, so the other points are positioned top of them, making them visible\n",
    "\n",
    "label <- umap_embed_rat_neurons %>%\n",
    "  group_by(celltype) %>%\n",
    "  summarize(\n",
    "    x = median(umap_1),\n",
    "    y = median(umap_2)\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "063d0012-8ff0-4107-a3f7-6d7b0055e597",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(umap_embed_rat_neurons, aes(x = umap_1, y = umap_2, color = color_group)) +\n",
    "  geom_point_rast(size = 0.1, alpha = 0.5) +\n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    axis.line = element_line(colour = \"black\", size = 0.4),\n",
    "    panel.grid = element_blank(),\n",
    "    panel.border = element_blank(),\n",
    "    panel.background = element_blank(),\n",
    "    legend.position = \"right\",\n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    axis.text = element_text(face = \"bold\")\n",
    "  ) +\n",
    "  scale_color_manual(values = c(\"rat\" = \"#546577\", \"Not expressed\" = \"grey\"), name = \"Expression\")\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4g_prlh.pdf\", width = 5, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d24af77-aa44-48b3-9721-45f68d20c052",
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_neurons <- subset(neurons, dataset == \"rat.2023\")\n",
    "DefaultAssay(rat_neurons) <- \"RNA\"\n",
    "\n",
    "rat_calcr_cells <- WhichCells(rat_neurons, expression = Calcr > 5)\n",
    "\n",
    "umap_embed_rat_neurons <- neurons@reductions$umap@cell.embeddings %>%\n",
    "  as.data.frame() %>%\n",
    "  mutate(\n",
    "    species = neurons$species,\n",
    "    celltype = neurons$cell.type,\n",
    "    Calcr_expressed = rownames(.) %in% rat_calcr_cells,  # Label cells where Calcr is expressed\n",
    "    color_group = ifelse(Calcr_expressed, as.character(species), \"Not expressed\"),\n",
    "    color_group = factor(color_group, levels = c(\"rat\", \"Not expressed\"))\n",
    "  ) %>%\n",
    "  sample_frac(1) %>%  # Shuffle rows\n",
    "  arrange(desc(color_group == \"Not expressed\"))  # Ensure background is plotted first, so the other points are positioned top of them, making them visible\n",
    "\n",
    "label <- umap_embed_rat_neurons %>%\n",
    "  group_by(celltype) %>%\n",
    "  summarize(\n",
    "    x = median(umap_1),\n",
    "    y = median(umap_2)\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8a7ddd-e6a7-4f40-bfce-e8cbfe135e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(umap_embed_rat_neurons, aes(x = umap_1, y = umap_2, color = color_group)) +\n",
    "  geom_point_rast(size = 0.1, alpha = 0.5) +\n",
    "  theme_pubr(base_size = 6, base_family = \"sans\") +\n",
    "  theme(\n",
    "    axis.line = element_line(colour = \"black\", size = 0.4),\n",
    "    panel.grid = element_blank(),\n",
    "    panel.border = element_blank(),\n",
    "    panel.background = element_blank(),\n",
    "    legend.position = \"right\",\n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    axis.text = element_text(face = \"bold\")\n",
    "  ) +\n",
    "  scale_color_manual(values = c(\"rat\" = \"#546577\", \"Not expressed\" = \"grey\"), name = \"Expression\")\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4g_calcr.pdf\", width = 5, height = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d57056-7f66-4349-a340-16d072e049f0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## H. Prlh expression in rat (single-cell pseudobulk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9194975b-5863-4dd6-9e9c-37ad7f99cd16",
   "metadata": {},
   "outputs": [],
   "source": [
    "neurons_sub <- neurons %>%\n",
    "  subset(\n",
    "    species == \"rat\" &\n",
    "      cell.type == \"Glu8.2\" &\n",
    "      treatment %in% c(\"A8-A\", \"A8-C\", \"V-A\", \"WM-C\")\n",
    "  ) %>%\n",
    "  subset(\n",
    "    hash.ID %in% names(which(table(hash.ID) >= 5))  # Remove samples with fewer than 5 cells\n",
    "  )\n",
    "DefaultAssay(neurons_sub) <- \"RNA\"\n",
    "\n",
    "genes <- rownames(neurons_sub)[rowMeans(neurons_sub@assays$RNA@counts != 0) >= 0.1]  # Get genes expressed in at least 10% of the cells\n",
    "samples <- unique(neurons_sub$hash.ID)\n",
    "\n",
    "# Prepare pseudobulk data\n",
    "pseudo_data <- data.frame(matrix(NA, nrow = length(genes), ncol = length(samples)))\n",
    "rownames(pseudo_data) <- genes\n",
    "colnames(pseudo_data) <- samples\n",
    "\n",
    "pseudo_meta <- data.frame(\n",
    "  sample = samples,\n",
    "  treatment = NA,\n",
    "  stringsAsFactors = FALSE\n",
    ")\n",
    "\n",
    "# Compute pseudobulk data using a for-loop\n",
    "for (j in seq_along(samples)) {\n",
    "  sample_cells <- which(neurons_sub$hash.ID == samples[j])\n",
    "  pseudo_data[, j] <- rowMeans(neurons_sub@assays$RNA@counts[genes, sample_cells, drop = FALSE])\n",
    "  pseudo_meta$treatment[j] <- unique(neurons_sub$treatment[sample_cells])\n",
    "}\n",
    "\n",
    "# Prepare expression data\n",
    "expr_data <- cbind.data.frame(t(pseudo_data), treatment = factor(pseudo_meta$treatment))\n",
    "expr_data <- melt(expr_data, id.vars = \"treatment\")\n",
    "colnames(expr_data) <- c(\"treatment\", \"gene\", \"expr\")\n",
    "\n",
    "expr_data$time <- \"4 hours\"\n",
    "expr_data$time[grep(\"-C\", expr_data$treatment)] <- \"8 days\"\n",
    "expr_data$drug <- \"Vehicle\"\n",
    "expr_data$drug[grep(\"A8-\", expr_data$treatment)] <- \"Cagrilintide\"\n",
    "\n",
    "# Prepare DGE results\n",
    "prlh_expr <- subset(expr_data, gene == \"Prlh\")\n",
    "prlh_expr$drug <- factor(prlh_expr$drug, levels = c(\"Vehicle\", \"Cagrilintide\"))\n",
    "p_values <- c(\n",
    "  DEGs.sc.rat.acute[[\"Glu8.2\"]][\"Prlh\", \"pvalue\"],\n",
    "  DEGs.sc.rat.chronic[[\"Glu8.2\"]][\"Prlh\", \"pvalue\"]\n",
    ")\n",
    "p_values <- round(p_values, 3)\n",
    "p_symbol <- ifelse(p_values <= 0.05, p_values, \"NS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e26ec969-277c-4425-9007-895090d1f3ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(prlh_expr, aes(x = time, y = expr, fill = drug)) +\n",
    "  geom_boxplot(\n",
    "    outlier.shape = NA,\n",
    "    position = position_dodge(0.6),\n",
    "    width = 0.5\n",
    "  ) +\n",
    "  theme_pubr(legend = \"top\") +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1, size = 6, face = \"bold\"),\n",
    "    axis.text.y = element_text(size = 6, face = \"bold\"),\n",
    "    axis.title.y = element_text(size = 6, face = \"bold\"),\n",
    "    plot.title = element_text(hjust = 0.5, size = 8, face = \"bold.italic\"),\n",
    "    legend.text = element_text(size = 6, face = \"bold\"),\n",
    "    legend.title = element_blank()\n",
    "  ) +\n",
    "  scale_fill_manual(values = c(\"#AFC3E4\", \"#95567D\")) + # Unsure about colors, may change later in illustrator\n",
    "  labs(\n",
    "    y = \"Pseudobulk expression\",\n",
    "    x = \"\",\n",
    "    title = \"Prlh\"\n",
    "  ) +\n",
    "  coord_cartesian(ylim = c(0, 15)) +\n",
    "  geom_signif(\n",
    "    y_position = c(9, 14),\n",
    "    xmin = c(0.85, 1.85),\n",
    "    xmax = c(1.15, 2.15),\n",
    "    annotation = p_symbol,\n",
    "    tip_length = 0.01,\n",
    "    textsize = 2,\n",
    "    size = 0.2\n",
    "  )\n",
    "\n",
    "p\n",
    "\n",
    "# ggsave(\"/projects/perslab/people/jmg776/projects/DVC/figures/2025/figure_4h.pdf\", width = 5, height = 4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.3.2",
   "language": "R",
   "name": "ir432"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
