{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7c2311b-15b2-4fee-98fa-a35f9d7c3b4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(rtracklayer)\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(ggpubr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6ad4782-99d0-4f64-9c99-33d29d4eaa67",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Mapping data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ec96717-2fd4-4760-8045-564bf9195548",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf.mouse <- import(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/reference/gencode.vM23.annotation.gtf\")\n",
    "gtf.mouse <- as.data.frame(gtf.mouse)\n",
    "\n",
    "gtf.rat <- import(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/reference/Rattus_norvegicus.Rnor_6.0.104.gtf\")\n",
    "gtf.rat <- as.data.frame(gtf.rat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "097398c6-bb42-4fbe-b514-0019284a61bb",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Process bulk transcriptomics data from mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8c76c4f-847f-427c-abd8-bca35f30685d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read bulk data\n",
    "main.dir <- \"/projects/mludwig/DVC/data/bulk_rna-seq/raw/out/mouse/\"\n",
    "samples <- list.dirs(main.dir, recursive = F, full.names = F) \n",
    "suffix <- \"ReadsPerGene.out.tab\"\n",
    "full.path <- paste0(main.dir, samples, \"/\", samples, suffix)\n",
    "\n",
    "counts.list <- lapply(full.path, function(x) {read.csv(x, sep = \"\", skip = 4, header = F,\n",
    "                                                       colClasses = c(\"character\", \"numeric\", \"NULL\", \"NULL\"),\n",
    "                                                       row.names = 1)})\n",
    "\n",
    "counts.mouse <- do.call(cbind, counts.list)\n",
    "names <- gsub(\"MGE01_|Grp.*\", \"\", samples)\n",
    "colnames(counts.mouse) <- names\n",
    "counts.mouse <- counts.mouse[, paste0(\"mouse\", 1:70)]\n",
    "\n",
    "# Remove genes with zero variance\n",
    "idx <- which(apply(counts.mouse, 1, var) != 0)\n",
    "counts.mouse <- counts.mouse[idx, ] \n",
    "\n",
    "## Map to gene names\n",
    "gene.names <- gtf.mouse$gene_name[match(rownames(counts.mouse), gtf.mouse$gene_id)]\n",
    "\n",
    "# Keep gene id if unique mapping is not possible\n",
    "keep.idx <- which(gene.names %in% unique(gene.names[duplicated(gene.names)]))\n",
    "rownames(counts.mouse)[-keep.idx] <- gene.names[-keep.idx]\n",
    "\n",
    "# Save\n",
    "write.csv(counts.mouse, file = \"/projects/mludwig/DVC/data/bulk_rna-seq/processed/counts_mouse.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a1db02d-21ff-406e-9cbc-832c144fb764",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load treatment groups\n",
    "groups.mouse <- openxlsx::read.xlsx(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/groups_mouse.xlsx\")\n",
    "groups.mouse <- groups.mouse[, c(1, 3)]\n",
    "colnames(groups.mouse) <- c(\"sample\", \"treatment\")\n",
    "groups.mouse$sample <- paste0(\"mouse\", groups.mouse$sample)\n",
    "groups.mouse$treatment <- gsub(1, \"V-A\", groups.mouse$treatment)\n",
    "groups.mouse$treatment <- gsub(2, \"A8-A\", groups.mouse$treatment)\n",
    "groups.mouse$treatment <- gsub(3, \"A8-C\", groups.mouse$treatment)\n",
    "groups.mouse$treatment <- gsub(4, \"WM-C\", groups.mouse$treatment)\n",
    "groups.mouse$treatment <- gsub(5, \"L-A\", groups.mouse$treatment)\n",
    "groups.mouse$treatment <- as.factor(groups.mouse$treatment)\n",
    "\n",
    "# Load QC data from different sequencing batches\n",
    "main.dir <- \"/projects/mludwig/DVC/data/bulk_rna-seq/raw/out/mouse/\"\n",
    "samples <- list.dirs(main.dir, recursive = F, full.names = F) \n",
    "suffix <- \"Log.final.out\"\n",
    "\n",
    "mapped.read.path <- paste0(main.dir, samples, \"/\", samples, suffix)\n",
    "\n",
    "QC.mouse1 <- data.frame(matrix(NA, nrow = 70, ncol = 2))\n",
    "colnames(QC.mouse1) <- c(\"sample\", \"mapped.reads\")\n",
    "for (i in 1:length(samples)) {\n",
    "  path <-  paste0(main.dir, samples[i], \"/\", samples[i], suffix)\n",
    "  mapped.reads <- read.csv(path, header = F, colClasses = c(NA, \"NULL\"))[9,]\n",
    "  mapped.reads <- as.numeric(gsub(\".*Uniquely mapped reads % \\\\|\\t|%$\", \"\", mapped.reads))\n",
    "  QC.mouse1$sample[i] <- samples[i]\n",
    "  QC.mouse1$mapped.reads[i] <- mapped.reads\n",
    "}\n",
    "QC.mouse1$sample <- gsub(\"MGE01_|Grp.*\", \"\", QC.mouse1$sample)\n",
    "\n",
    "QC.mouse2 <- read.csv(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/90-612232719_statistics.csv\")\n",
    "QC.mouse2 <- QC.mouse2[grep(\"mouse\", QC.mouse2$Sample.ID), ]\n",
    "colnames(QC.mouse2) <- c(\"project\", \"sample\", \"barcode\", \"reads\", \"yield\", \"quality.score\", \"bases\")\n",
    "QC.mouse2 <- QC.mouse2[, c(\"sample\", \"reads\", \"yield\", \"quality.score\", \"bases\")]\n",
    "QC.mouse2$sample <- gsub(\"MGE01_|Grp.*\", \"\", QC.mouse2$sample)\n",
    "QC.mouse2$reads <- as.numeric(gsub(\"\\\\,\", \"\", QC.mouse2$reads))\n",
    "QC.mouse2$yield <- as.numeric(gsub(\"\\\\,\", \"\", QC.mouse2$yield))\n",
    "\n",
    "QC.mouse3 <- read.csv(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/90-612232719_rRNA.txt\", sep = \"\\t\")\n",
    "QC.mouse3 <- QC.mouse3[grep(\"mouse\", QC.mouse3$Sample.ID), ]\n",
    "colnames(QC.mouse3) <- c(\"sample\", \"rRNA\")\n",
    "QC.mouse3$sample <- gsub(\"MGE01_|Grp.*\", \"\", QC.mouse3$sample)\n",
    "QC.mouse3$rRNA <- as.numeric(gsub(\"\\\\,\", \"\\\\.\", QC.mouse3$rRNA))\n",
    "\n",
    "QC.mouse <- merge(QC.mouse1, QC.mouse2, by = \"sample\")\n",
    "QC.mouse <- merge(QC.mouse, QC.mouse3, by = \"sample\")\n",
    "QC.mouse <- QC.mouse[match(paste0(\"mouse\", 1:70), QC.mouse$sample),]\n",
    "QC.mouse\n",
    "\n",
    "\n",
    "# Merge\n",
    "meta.mouse <- merge(x = groups.mouse, y = QC.mouse, by = \"sample\")\n",
    "meta.mouse <- meta.mouse[match(groups.mouse$sample, meta.mouse$sample), ]\n",
    "\n",
    "# Save\n",
    "write.csv(meta.mouse, file = \"/projects/mludwig/DVC/data/bulk_rna-seq/processed/meta_mouse.csv\", quote = F, row.names = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "206151a9-3f87-4997-9964-126baa6b15a1",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Process bulk transcriptomics data from rat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f8a9d3-8fc2-430f-938a-d3216e712cad",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Read bulk data\n",
    "main.dir <-  \"/projects/mludwig/DVC/data/bulk_rna-seq/raw/out/rat/\"\n",
    "samples <- list.dirs(main.dir, recursive = F, full.names = F) \n",
    "suffix <- \"ReadsPerGene.out.tab\"\n",
    "full.path <- paste0(main.dir, samples, \"/\", samples, suffix)\n",
    "\n",
    "\n",
    "counts.list <- lapply(full.path, function(x) {read.csv(x, sep = \"\", skip = 4, header = F,\n",
    "                                                       colClasses = c(\"character\", \"numeric\", \"NULL\", \"NULL\"),\n",
    "                                                       row.names = 1)})\n",
    "\n",
    "counts.rat <- do.call(cbind, counts.list)\n",
    "names <-  gsub(\"MGE01_|Grp.*\", \"\", samples)\n",
    "names <- gsub(\"Rat\", \"rat\", names)\n",
    "colnames(counts.rat) <- names \n",
    "counts.rat <- counts.rat[, paste0(\"rat\", 1:70)]\n",
    "\n",
    "# Remove genes with zero variance\n",
    "idx <- which(apply(counts.rat, 1, var) != 0)\n",
    "counts.rat <- counts.rat[idx, ] \n",
    "\n",
    "## Map to gene names\n",
    "gene.names <- gtf.rat$gene_name[match(rownames(counts.rat), gtf.rat$gene_id)]\n",
    "\n",
    "# Keep gene id if unique mapping is not possible\n",
    "keep.idx <- which(gene.names %in% unique(gene.names[duplicated(gene.names)]))\n",
    "\n",
    "rownames(counts.rat)[-keep.idx] <- gene.names[-keep.idx]\n",
    "\n",
    "# Save\n",
    "write.csv(counts.rat, file = \"/projects/mludwig/DVC/data/bulk_rna-seq/processed/counts_rat.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f33910c9-b5ab-48aa-b0b6-c6151c0ed883",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Load treatment groups\n",
    "groups.rat <- openxlsx::read.xlsx(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/groups_rat.xlsx\")\n",
    "colnames(groups.rat) <- c(\"sample\", \"treatment\")\n",
    "groups.rat$sample <- paste0(\"rat\", groups.rat$sample)\n",
    "groups.rat$treatment <- gsub(1, \"V-A\", groups.rat$treatment)\n",
    "groups.rat$treatment <- gsub(2, \"A8-A\", groups.rat$treatment)\n",
    "groups.rat$treatment <- gsub(3, \"A8-C\", groups.rat$treatment)\n",
    "groups.rat$treatment <- gsub(4, \"WM-C\", groups.rat$treatment)\n",
    "groups.rat$treatment <- gsub(5, \"L-A\", groups.rat$treatment)\n",
    "\n",
    "\n",
    "# Load QC data from different sequencing batches\n",
    "main.dir <- \"/projects/mludwig/DVC/data/bulk_rna-seq/raw/out/rat/\"\n",
    "samples <- list.dirs(main.dir, recursive = F, full.names = F) \n",
    "suffix <- \"Log.final.out\"\n",
    "\n",
    "mapped.read.path <- paste0(main.dir, samples, \"/\", samples, suffix)\n",
    "\n",
    "QC.rat1 <- data.frame(matrix(NA, nrow = 70, ncol = 2))\n",
    "colnames(QC.rat1) <- c(\"sample\", \"mapped.reads\")\n",
    "for (i in 1:length(samples)) {\n",
    "  path <-  paste0(main.dir, samples[i], \"/\", samples[i], suffix)\n",
    "  mapped.reads <- read.csv(path, header = F, colClasses = c(NA, \"NULL\"))[9,]\n",
    "  mapped.reads <- as.numeric(gsub(\".*Uniquely mapped reads % \\\\|\\t|%$\", \"\", mapped.reads))\n",
    "  QC.rat1$sample[i] <- samples[i]\n",
    "  QC.rat1$mapped.reads[i] <- mapped.reads\n",
    "}\n",
    "QC.rat1$sample <- gsub(\"MGE01_|Grp.*\", \"\", QC.rat1$sample)\n",
    "QC.rat1$sample <- gsub(\"Rat\", \"rat\", QC.rat1$sample)\n",
    "\n",
    "QC.rat2 <- read.csv(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/90-612232719_statistics.csv\")\n",
    "QC.rat2 <- QC.rat2[grep(\"Rat\", QC.rat2$Sample.ID), ]\n",
    "colnames(QC.rat2) <- c(\"project\", \"sample\", \"barcode\", \"reads\", \"yield\", \"quality.score\", \"bases\")\n",
    "QC.rat2 <- QC.rat2[, c(\"sample\", \"reads\", \"yield\", \"quality.score\", \"bases\")]\n",
    "QC.rat2$sample <- gsub(\"MGE01_|Grp.*\", \"\", QC.rat2$sample)\n",
    "QC.rat2$sample <- gsub(\"Rat\", \"rat\", QC.rat2$sample)\n",
    "QC.rat2$reads <- as.numeric(gsub(\"\\\\,\", \"\", QC.rat2$reads))\n",
    "QC.rat2$yield <- as.numeric(gsub(\"\\\\,\", \"\", QC.rat2$yield))\n",
    "\n",
    "QC.rat3 <- read.csv(\"/projects/mludwig/DVC/data/bulk_rna-seq/raw/90-612232719_rRNA.txt\", sep = \"\\t\")\n",
    "QC.rat3 <- QC.rat3[grep(\"Rat\", QC.rat3$Sample.ID), ]\n",
    "colnames(QC.rat3) <- c(\"sample\", \"rRNA\")\n",
    "QC.rat3$sample <- gsub(\"MGE01_|Grp.*\", \"\", QC.rat3$sample)\n",
    "QC.rat3$sample <- gsub(\"Rat\", \"rat\", QC.rat3$sample)\n",
    "QC.rat3$rRNA <- as.numeric(gsub(\"\\\\,\", \"\\\\.\", QC.rat3$rRNA))\n",
    "\n",
    "QC.rat <- merge(QC.rat1, QC.rat2, by = \"sample\")\n",
    "QC.rat <- merge(QC.rat, QC.rat3, by = \"sample\")\n",
    "QC.rat <- QC.rat[match(paste0(\"rat\", 1:70), QC.rat$sample),]\n",
    "QC.rat\n",
    "\n",
    "\n",
    "# Merge\n",
    "meta.rat <- merge(x = groups.rat, y = QC.rat, by = \"sample\")\n",
    "meta.rat <- meta.rat[match(groups.rat$sample, meta.rat$sample), ]\n",
    "\n",
    "# Save\n",
    "write.csv(meta.rat, file = \"/projects/mludwig/DVC/data/bulk_rna-seq/processed/meta_rat.csv\", quote = F, row.names = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a1418cd-8733-4a1a-aa39-39c5829b4c01",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Identify outliers in mouse transcriptomics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c18e4c-afc1-451f-b045-43fe783f9f1c",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "design <- data.frame(condition = meta.mouse$treatment, rRNA = meta.mouse$rRNA)\n",
    "dds <- DESeqDataSetFromMatrix(counts.mouse, DataFrame(design), ~ condition + rRNA)\n",
    "dds <- DESeq(dds)\n",
    "vsd <- vst(dds, blind=F)\n",
    "\n",
    "pca.data <- plotPCA(vsd, intgroup=c(\"condition\"), returnData=TRUE, ntop = nrow(assay(vsd)))\n",
    "percent.var <- round(100 * attr(pca.data, \"percentVar\"))\n",
    "pca.data$condition <- factor(pca.data$condition)\n",
    "pca.data$rRNA <- meta.mouse$rRNA\n",
    "\n",
    "label <- data.frame(condition = unique(pca.data$condition))\n",
    "\n",
    "label <- pca.data %>% \n",
    "  dplyr::group_by(condition) %>% \n",
    "  dplyr::summarize(x = median(PC1), y = median(PC2)) \n",
    "\n",
    "pca.plot <- ggplot(pca.data, aes(PC1, PC2, color = rRNA)) +\n",
    "  geom_point(size = 2) +\n",
    "  xlab(paste0(\"PC1 (\", percent.var[1], \"% variance)\")) +\n",
    "  ylab(paste0(\"PC2 (\", percent.var[2], \"% variance)\")) + \n",
    "  theme_pubr() + theme(panel.border = element_blank(), \n",
    "                     panel.grid.major = element_blank(),\n",
    "                     panel.grid.minor = element_blank(), \n",
    "                     axis.line = element_line(colour = \"black\"),\n",
    "                     axis.text = element_text(hjust = 1, face = \"bold\"),\n",
    "                     axis.title = element_text(size=10, face = \"bold\"),\n",
    "                     legend.title = element_blank()) +   \n",
    "  scale_x_continuous(breaks = seq(-500, 500, 10)) \n",
    "pca.plot\n",
    "\n",
    "# Remove outliers with > 3 SD on PC1 and PC2\n",
    "pca.data2 <- prcomp(t(assay(vsd)))\n",
    "head(apply(pca.data2$x, 2, function(x) which( abs(x - mean(x)) > (3 * sd(x)))), n = 2)\n",
    "\n",
    "\n",
    "# Mouse11 and mouse33 are identified as outliers and are removed (first iteration)\n",
    "# Mouse25 and mouse27 are identified as outliers and are removed (second iteration)\n",
    "samples.sub <- colnames(counts.mouse)[!(colnames(counts.mouse) %in% c(\"mouse11\", \"mouse25\", \"mouse27\", \"mouse33\"))]\n",
    "counts.sub <- counts.mouse[, samples.sub]\n",
    "meta.sub <- meta.mouse[match(samples.sub, meta.mouse$sample), ]\n",
    "\n",
    "design <- data.frame(condition = meta.sub$treatment, rRNA = meta.sub$rRNA)\n",
    "dds <- DESeqDataSetFromMatrix(counts.sub, DataFrame(design), ~ condition + rRNA)\n",
    "dds <- DESeq(dds)\n",
    "vsd <- vst(dds, blind=F)\n",
    "\n",
    "pca.data <- plotPCA(vsd, intgroup=c(\"condition\"), returnData=TRUE, ntop = nrow(assay(vsd)))\n",
    "percent.var <- round(100 * attr(pca.data, \"percentVar\"))\n",
    "pca.data$condition <- factor(pca.data$condition)\n",
    "pca.data$rRNA <- meta.sub$rRNA\n",
    "\n",
    "label <- data.frame(condition = unique(pca.data$condition))\n",
    "\n",
    "label <- pca.data %>% \n",
    "  dplyr::group_by(condition) %>% \n",
    "  dplyr::summarize(x = median(PC1), y = median(PC2)) \n",
    "\n",
    "pca.plot <- ggplot(pca.data, aes(PC1, PC2, color = rRNA)) +\n",
    "  geom_point(size = 2) +\n",
    "  xlab(paste0(\"PC1 (\", percent.var[1], \"% variance)\")) +\n",
    "  ylab(paste0(\"PC2 (\", percent.var[2], \"% variance)\")) + \n",
    "  theme_pubr() + theme(panel.border = element_blank(), \n",
    "                     panel.grid.major = element_blank(),\n",
    "                     panel.grid.minor = element_blank(), \n",
    "                     axis.line = element_line(colour = \"black\"),\n",
    "                     axis.text = element_text(hjust = 1, face = \"bold\"),\n",
    "                     axis.title = element_text(size=10, face = \"bold\"),\n",
    "                     legend.title = element_blank()) +   \n",
    "  scale_x_continuous(breaks = seq(-500, 500, 10)) \n",
    "pca.plot\n",
    "\n",
    "# Remove outliers with > 3 SD on PC1 and PC2\n",
    "pca.data2 <- prcomp(t(assay(vsd)))\n",
    "head(apply(pca.data2$x, 2, function(x) which( abs(x - mean(x)) > (3 * sd(x)))), n = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b485fad-96af-4511-8ca6-8a3ba782a1db",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Identify outliers in rat transcriptomics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24b7bd2e-f00f-4783-af13-97f3258016e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "design <- data.frame(condition = meta.rat$treatment, rRNA = meta.rat$rRNA)\n",
    "dds <- DESeqDataSetFromMatrix(counts.rat, DataFrame(design), ~ condition + rRNA)\n",
    "dds <- DESeq(dds)\n",
    "vsd <- vst(dds, blind=F)\n",
    "\n",
    "pca.data <- plotPCA(vsd, intgroup=c(\"condition\"), returnData=TRUE, ntop = nrow(assay(vsd)))\n",
    "percent.var <- round(100 * attr(pca.data, \"percentVar\"))\n",
    "pca.data$condition <- factor(pca.data$condition)\n",
    "pca.data$rRNA <- meta.rat$rRNA\n",
    "\n",
    "label <- data.frame(condition = unique(pca.data$condition))\n",
    "\n",
    "label <- pca.data %>% \n",
    "  dplyr::group_by(condition) %>% \n",
    "  dplyr::summarize(x = median(PC1), y = median(PC2)) \n",
    "\n",
    "pca.plot <- ggplot(pca.data, aes(PC1, PC2, color = rRNA)) +\n",
    "  geom_point(size = 2) +\n",
    "  xlab(paste0(\"PC1 (\", percent.var[1], \"% variance)\")) +\n",
    "  ylab(paste0(\"PC2 (\", percent.var[2], \"% variance)\")) + \n",
    "  theme_pubr() + theme(panel.border = element_blank(), \n",
    "                     panel.grid.major = element_blank(),\n",
    "                     panel.grid.minor = element_blank(), \n",
    "                     axis.line = element_line(colour = \"black\"),\n",
    "                     axis.text = element_text(hjust = 1, face = \"bold\"),\n",
    "                     axis.title = element_text(size=10, face = \"bold\"),\n",
    "                     legend.title = element_blank()) +   \n",
    "  scale_x_continuous(breaks = seq(-500, 500, 10)) \n",
    "pca.plot\n",
    "\n",
    "# Remove outliers with > 3 SD on PC1 and PC2\n",
    "pca.data2 <- prcomp(t(assay(vsd)))\n",
    "head(apply(pca.data2$x, 2, function(x) which( abs(x - mean(x)) > (3 * sd(x)))), n = 2)\n",
    "\n",
    "\n",
    "# Rat9 and rat65 are identified as outliers and are removed\n",
    "samples.sub <- colnames(counts.rat)[!(colnames(counts.rat) %in% c(\"rat9\", \"rat65\"))]\n",
    "counts.sub <- counts.rat[, samples.sub]\n",
    "meta.sub <- meta.rat[meta.rat$sample %in% samples.sub, ]\n",
    "\n",
    "design <- data.frame(condition = meta.sub$treatment, rRNA = meta.sub$rRNA)\n",
    "dds <- DESeqDataSetFromMatrix(counts.sub, DataFrame(design), ~ condition + rRNA)\n",
    "dds <- DESeq(dds)\n",
    "vsd <- vst(dds, blind=F)\n",
    "\n",
    "pca.data <- plotPCA(vsd, intgroup=c(\"condition\"), returnData=TRUE, ntop = nrow(assay(vsd)))\n",
    "percent.var <- round(100 * attr(pca.data, \"percentVar\"))\n",
    "pca.data$condition <- factor(pca.data$condition)\n",
    "pca.data$rRNA <- meta.sub$rRNA\n",
    "\n",
    "label <- data.frame(condition = unique(pca.data$condition))\n",
    "\n",
    "label <- pca.data %>% \n",
    "  dplyr::group_by(condition) %>% \n",
    "  dplyr::summarize(x = median(PC1), y = median(PC2)) \n",
    "\n",
    "pca.plot <- ggplot(pca.data, aes(PC1, PC2, color = rRNA)) +\n",
    "  geom_point(size = 2) +\n",
    "  xlab(paste0(\"PC1 (\", percent.var[1], \"% variance)\")) +\n",
    "  ylab(paste0(\"PC2 (\", percent.var[2], \"% variance)\")) + \n",
    "  theme_pubr() + theme(panel.border = element_blank(), \n",
    "                     panel.grid.major = element_blank(),\n",
    "                     panel.grid.minor = element_blank(), \n",
    "                     axis.line = element_line(colour = \"black\"),\n",
    "                     axis.text = element_text(hjust = 1, face = \"bold\"),\n",
    "                     axis.title = element_text(size=10, face = \"bold\"),\n",
    "                     legend.title = element_blank()) +   \n",
    "  scale_x_continuous(breaks = seq(-500, 500, 10)) \n",
    "pca.plot\n",
    "\n",
    "# No more outliers are detected\n",
    "pca.data2 <- prcomp(t(assay(vsd)))\n",
    "head(apply(pca.data2$x, 2, function(x) which( abs(x - mean(x)) > (3 * sd(x)))), n = 2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_py3.11",
   "language": "python",
   "name": "conda_py3.11"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
